### PR TITLE
Turn core code into a PyPI library

### LINKED ISSUE #556: Decouple iris core and package into pypi
- [ ] Decouple the iris core functionality from its test harness and control center. Should be able to function as completely separate entities. 

- [ ] Create a pip package and upload to pypi
- [ ] Update documentation to reflect these changes

### PR DESCRIPTION
**Pull request by mwobensmith**
Fixes #556 

Notes:

* The tests/target for 'sample' will exist in the repo, but will not be part of the package.

* I am performing this PR against `master` and not `dev` because `dev` has already diverged from when I did my initial code pull. There will likely be some minor code updates to get late-breaking changes into this library before it goes live.

* Aside from command `iris` (as before), there is another command `api-test` that will show you a basic example of moving the mouse on screen. This code only invokes the test API and nothing else in the project.

I'm sure there will be more! ;)

### CODE REVIEW

File: moziris/api/finder/text_search.py [hunk 1]
```diff
@@ -134,7 +134,7 @@ def _text_search(text, region: Rectangle = None, multiple_search=False):
         first_word = first_word_occurrences
 
     if word_count == 1:
-        save_debug_ocr_image(text, img, first_word)
+        #save_debug_ocr_image(text, img, first_word)
```
Comment (kimberlythegeek) [inline moziris/api/finder/text_search.py]: mind removing the comment if this is no longer needed?

File: moziris/email_report/email_client.py [hunk 2]
```diff
@@ -51,22 +50,18 @@ def get_file_attachment():
...
-                    repo_details.get('iris_branch_head'), 
-                    test_status)
-                )) 
+            body_message = MIMEText(
+                ''' Repo_details:\n ''' + """Branch_name:""" + repo_details.get(
+                    'iris_branch')
+                + " \n " + '''Branch_head: ''' + repo_details.get(
+                    'iris_branch_head') + "\n\n" + '''Test_Run_Details: ''' + test_status +
+                ''' \nNote: To see the complete run output, please check the attachment.''')
         else:
```
Comment (kimberlythegeek) [inline moziris/email_report/email_client.py]: What's reason for the change in string interpolation? The original code is using the new implementation for python 3

**Comment by kimberlythegeek** [comment]
I was trying to run this against the firefox tests, following https://github.com/mwobensmith/iristemp/wiki/Setup-instructions

the `hi` and `api` commands worked, but the `iris` commands all appeared to hang, I let them run for several minutes and they printed no ouput.

**Comment by tracywalker** [comment]
> I was trying to run this against the firefox tests, following https://github.com/mwobensmith/iristemp/wiki/Setup-instructions
> 
> the `hi` and `api` commands worked, but the `iris` commands all appeared to hang, I let them run for several minutes and they printed no ouput.

I originally had the same issue.  Turns out I had not properly done the 
```
$ export PYTHONPATH=$PWD
$ export IRIS_CODE_ROOT=$PWD
```
That must be done in the pipenv shell

**Comment by kimberlythegeek** [comment]
@tracywalker I tried that and ran `iris sample` again. stills hangs

File: moziris/email_report/email_client.py (hunk 2 above)
Comment (mwobensmith) [reply moziris/email_report/email_client.py]: Good catch. My code went out of sync with the master repo during the rewrite, and old code crept back in.

**Comment by mwobensmith** [comment]
I have addressed the code-related concerns.

Regarding the bootstrap script, it's a requirement for people right now. Are you proposing that we require people to move to Docker in order to use our project?

**Comment by mwobensmith** [comment]
> @tracywalker I tried that and ran `iris sample` again. stills hangs

What platform? Also, have you ever run Iris on this machine before?

**Comment by kimberlythegeek** [comment]
@tracywalker OSX. yes, I can run iris from master.

@mwobensmith an application shouldn't directly modify a user's python version without asking permission first. this should be done within a contained environment. tools like pipenv and tox can create virtual environments for specific versions of python. this prevents contamination or modification of a user's machine.

No, I'm not suggesting that docker be a requirement for anyone to use. but this should not be made publicly available (read: pypi.org) unless we make some changes to the bootstrap.

**Comment by mwobensmith** [comment]
@kimberlythegeek The bootstrap scripts are not part of the PyPI package and are not material to this discussion. We are in agreement about some of the points you raise there, but those should be brought up in a separate issue.

**Comment by kimberlythegeek** [comment]
@mwobensmith I guess I'm confused. How do you install and run this as a package without using the bootstrap file?

**Comment by kimberlythegeek** [comment]
@mwobensmith to clarify, I'm not saying the bootstrap is a reason to block this merge to master. 

My only objection is uploading it to the public pypi server while it operates in this manner. Seeing as our audience is internal, there's no reason to upload to pypi.org just yet.

**Comment by kimberlythegeek** [comment]
as far as the code, it all looks pretty straightforward. I have not been able to run it locally, but as I understand it, the exalt team will be doing the testing for that. 

Still learning what the team's expectations are for a code review. On past projects, pulling and running the code was just part of the process.

If the testing is expected to be done by the contractors, then I will approve merging this PR 👌

**Comment by mwobensmith** [comment]
> @mwobensmith I guess I'm confused. How do you install and run this as a package without using the bootstrap file?

Our project is a Python project, but not 100% pure. It relies on some native libraries like Tesseract, for example, for which there is a Python interface. So, in order to assure that our project works, you're going to have to guarantee some system dependencies. Right now, the bootstrap ensures that, but moving forward, we'd like better solutions there.

However, independent of *how* you get your system configured - bootstrap, Docker, etc. - the PyPI package installs via pip like any other.  

In summary, going from what we have now to a PyPI package is just moving code around for better reuse. The current system of duplicate code is unmanageable. This is the problem we are trying to solve.

**Comment by mwobensmith** [comment]
> @mwobensmith to clarify, I'm not saying the bootstrap is a reason to block this merge to master.
> 
> My only objection is uploading it to the public pypi server while it operates in this manner. Seeing as our audience is internal, there's no reason to upload to pypi.org just yet.

In order for the `iris_firefox` project to use this, it will have to exist as a public PyPI library. There will then be some work in that repo to eliminate the duplicate code and to redirect the imports to the package instead.

**Comment by mwobensmith** [comment]
> as far as the code, it all looks pretty straightforward. I have not been able to run it locally, but as I understand it, the exalt team will be doing the testing for that.
> 
> Still learning what the team's expectations are for a code review. On past projects, pulling and running the code was just part of the process.
> 
> If the testing is expected to be done by the contractors, then I will approve merging this PR 👌

We definitely want to make running the code part of the review process, and I don't think you should approve this until you can see it running with your own eyes.   👀

I have to add the latest code that came in since I did the port, so my work here is not done. It shouldn't be much. When that lands, you can take a look at any changes, and we can also further troubleshoot if it still does not work for you. 

Regardless of that, we can have Exalt et al run a pass on this as well.

**Comment by kimberlythegeek** [comment]
@mwobensmith understood. Just don’t want to hold up the process unnecessarily 👌🏻

**Comment by mwobensmith** [comment]
Just added missing code. Please review and comment.

**Comment by kimberlythegeek** [comment]
I have pulled down the latest changes and verified that the code works as expected.

File: moziris/api/finder/pattern.py [hunk 3]
```diff
@@ -24,38 +24,50 @@
...
+
+    def load_pattern(self, path = None):
+        if self.loaded:
+            return
 
-        if from_path is None:
-            path = _get_image_path(inspect.stack()[1][1], image_name)
+        if path is None:
+            path = _get_image_path(self.caller, self.temp_name)
         else:
```
Comment (kimberlythegeek) [inline moziris/api/finder/pattern.py]: Just wondering, is this `else` actually necessary?

File: moziris/api/finder/pattern.py [hunk 4]
```diff
@@ -173,6 +198,7 @@ def _get_array_from_image(image: Image):
 
 
 def _get_pattern_size(image: Image, scale: float) -> (int, int):
+    """Returns a tuple with values for image width and height"""
```
Comment (kimberlythegeek) [inline moziris/api/finder/pattern.py]: +1 for comments 😄

**Comment by kimberlythegeek** [comment]
`Scripts/` should be removed from the `.gitignore` file so that moziris/scripts is included

**Comment by mwobensmith** [comment]
> generally looks straightforward. I will pull and run code to verify it works.
> 
> To clarify, are we waiting to merge until we get this working with iristemp?

Yes.

File: moziris/api/finder/pattern.py (hunk 3 above)
Comment (mwobensmith) [reply moziris/api/finder/pattern.py]: Yes, because this function can be invoked two different ways, resulting in `path` either being `None` or a predetermined path. If it's the former, we will look up the path on disk for the Pattern, but otherwise, we'll use the value that is passed in instead.

Comment (mwobensmith) [reply moziris/api/finder/pattern.py]: Actually, no it is not. Thank you for catching that. :)

**Comment by mwobensmith** [comment]
@kimberlythegeek 
I have addressed review comments. This should be ready to test again.

**Comment by mwobensmith** [comment]
This was merged as part of the work for #553.
